### PR TITLE
[Lang] Experimental SpMV and direct linear solvers

### PR DESCRIPTION
**Pull request by FantasyVR**
Related pr = https://github.com/taichi-dev/taichi/pull/2657 and https://github.com/taichi-dev/taichi/pull/2792

In this pr, sparse matrix-vector multiplication and direct linear solvers (LLT, LDLT, LU) are supported.

### CODE REVIEW

File: taichi/program/sparse_matrix.cpp [hunk 1]
```diff
@@ -100,6 +104,12 @@ SparseMatrix SparseMatrix::matmul(const SparseMatrix &sm) {
   return SparseMatrix(res);
 }
 
+Eigen::VectorXf SparseMatrix::mat_vec_mul(
+    const Eigen::Ref<const Eigen::VectorXf> &b) {
+  Eigen::VectorXf res = matrix_ * b;
+  return res;
```
Comment (yuanming-hu) [inline taichi/program/sparse_matrix.cpp:110]:
```suggestion
  return matrix_ * b;
```

File: taichi/program/sparse_solver.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#include "Eigen/Sparse"
+
+namespace taichi {
+namespace lang {
+
+class SparseLUSolver {
+ public:
+  SparseLUSolver();
+  bool compute(const SparseMatrix &sm);
+  void analyzePattern(const SparseMatrix &sm);
```
Comment (yuanming-hu) [inline taichi/program/sparse_solver.h:16]: Following our coding convention, the function name should be `analyze_pattern`.

File: python/taichi/lang/sparse_solver.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            elif solver_type == "LU":
+                self.solver = _ti_core.SparseLUSolver()
+        else:
+            assert False, f"The solver type is not support for now."
+
+    def compute(self, sparse_matrix):
+        if isinstance(sparse_matrix, SparseMatrix):
+            self.solver.compute(sparse_matrix.matrix)
+        else:
+            assert False, f"The parameter type: {type(sparse_matrix)} is not support in linear solver for now."
```
Comment (yuanming-hu) [inline python/taichi/lang/sparse_solver.py:23]: The error string appears four times. Following [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), you may want to extract it into a (static) function.

File: python/taichi/lang/sparse_matrix.py [hunk 4]
```diff
@@ -39,9 +39,15 @@ def transpose(self):
...
-        return SparseMatrix(sm=sm)
+        import numpy as np
+        if isinstance(other, SparseMatrix):
+            assert self.m == other.n, f"Dimension mismatch between sparse matrices ({self.n}, {self.m}) and ({other.n}, {other.m})"
+            sm = self.matrix.matmul(other.matrix)
+            return SparseMatrix(sm=sm)
+        elif isinstance(other, np.ndarray):
+            assert self.m == other.shape[
+                0], f"Dimension mismatch between sparse matrix ({self.n}, {self.m}) and vector ({other.shape})"
+            return self.matrix.mat_vec_mul(other)
```
Comment (yuanming-hu) [inline python/taichi/lang/sparse_matrix.py:50]: You may want to call `.to_numpy()` here so that your users don't have to when calling `A @ b`.

**Review by yuanming-hu — COMMENTED** [review]
Cool! Just some minor comments.

File: python/taichi/lang/sparse_matrix.py [hunk 5]
```diff
@@ -39,9 +39,18 @@ def transpose(self):
...
+        import numpy as np
+        from taichi.lang import Field
+        if isinstance(other, SparseMatrix):
+            assert self.m == other.n, f"Dimension mismatch between sparse matrices ({self.n}, {self.m}) and ({other.n}, {other.m})"
+            sm = self.matrix.matmul(other.matrix)
+            return SparseMatrix(sm=sm)
+        elif isinstance(other, Field):
+            assert self.m == other.shape[
+                0], f"Dimension mismatch between sparse matrix ({self.n}, {self.m}) and vector ({other.shape})"
+            return self.matrix.mat_vec_mul(other.to_numpy())
```
Comment (yuanming-hu) [inline python/taichi/lang/sparse_matrix.py:51]: What if the user directly passes in a numpy array here?

File: python/taichi/lang/sparse_solver.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        if solver_type in solver_map:
+            from taichi.core.util import ti_core as _ti_core
+            if solver_type == "LLT":
+                self.solver = _ti_core.SparseLLTSolver()
+            elif solver_type == "LDLT":
+                self.solver = _ti_core.SparseLDLTSolver()
+            elif solver_type == "LU":
+                self.solver = _ti_core.SparseLUSolver()
+        else:
+            assert False, f"The solver type is not support for now."
```
Comment (yuanming-hu) [inline python/taichi/lang/sparse_solver.py:17]:
```suggestion
        from taichi.core.util import ti_core as _ti_core
        solver_map = {"LLT": lambda: _ti_core.SparseLLTSolver(), "LDLT": lambda: _ti_core.SparseLDLTSolver(), "LU": lambda: _ti_core.SparseLUSolver()}
        if solver_type in solver_map:
            self.solver = solver_map[solver_type]()
        else:
            assert False, f"The solver type '{solver_type}' is not supported for now. Supported types are {solver_map.keys()}."
```

File: tests/python/test_sparse_linear_solver.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            b[i] = i + 1
+
+    fill(Abuilder, Aarray, b)
+    A = Abuilder.build()
+    solver = ti.SparseSolver(solver_type="LU")
+    solver.analyze_pattern(A)
+    solver.factorize(A)
+    x = solver.solve(b)
+    for i in range(n):
+        assert x[i] - res[i] < 1.0e-6
```
Comment (yuanming-hu) [inline tests/python/test_sparse_linear_solver.py:100]: Consider `@pytest.mark.parametrize(...)`

File: taichi/python/export_lang.cpp [hunk 8]
```diff
@@ -979,6 +983,27 @@ void export_lang(py::module &m) {
...
+      .def("analyze_pattern", &SparseLDLTSolver::analyze_pattern)
+      .def("factorize", &SparseLDLTSolver::factorize)
+      .def("solve", &SparseLDLTSolver::solve);
+
+  py::class_<SparseLLTSolver>(m, "SparseLLTSolver")
+      .def(py::init<>())
+      .def("compute", &SparseLLTSolver::compute)
+      .def("analyze_pattern", &SparseLLTSolver::analyze_pattern)
+      .def("factorize", &SparseLLTSolver::factorize)
+      .def("solve", &SparseLLTSolver::solve);
```
Comment (yuanming-hu) [inline taichi/python/export_lang.cpp:1006]: Use a base class `SparseSolver`, and create 3 child classes `template <typename EigenSolver> EigenSparseSolver : public SparseSolver {private: EigenSolver solver_; ...}`.

Create a function method
```C++
std::unique_ptr<SparseSolver> get_sparse_solver(const std::string &name, Arch arch, DataType dtype) {
    TI_ASSERT(arch_is_cpu(arch));
    TI_ASSERT(dtype == DataType::f32);
    if (name == "LU") {
        return std::make_unique<EigenSparseSolver<LU>>();
    }
}
```

Comment (yuanming-hu) [reply taichi/python/export_lang.cpp:1006]: Or we leave this refactoring to the community (good first issue!) :-)

File: python/taichi/lang/sparse_matrix.py (hunk 5 above)
Comment (FantasyVR) [reply python/taichi/lang/sparse_matrix.py:51]: I see, I would add the np.ndarray as parameters.

File: taichi/python/export_lang.cpp (hunk 8 above)
Comment (FantasyVR) [reply taichi/python/export_lang.cpp]: I remove the repeated parts and update the sparse solver according to your advice. Thanks.

File: python/taichi/lang/sparse_matrix.py [hunk 9]
```diff
@@ -39,9 +39,22 @@ def transpose(self):
...
+        elif isinstance(other, np.ndarray):
+            assert self.m == other.shape[
+                0], f"Dimension mismatch between sparse matrix ({self.n}, {self.m}) and vector ({other.shape})"
+            return self.matrix.mat_vec_mul(other)
+        else:
+            assert False, f"Sparse matrix-matrix/vector multiplication does not support {type(other)} for now. Supported types are SparseMatrix, Field, numpy.ndarray."
```
Comment (yuanming-hu) [inline python/taichi/lang/sparse_matrix.py:57]:
```suggestion
            assert False, f"Sparse matrix-matrix/vector multiplication does not support {type(other)} for now. Supported types are SparseMatrix, ti.field, and numpy.ndarray."
```

File: python/taichi/lang/sparse_solver.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+    def solve(self, b):
+        import numpy as np
+        from taichi.lang import Field
+        if isinstance(b, Field):
+            return self.solver.solve(b.to_numpy())
+        elif isinstance(b, np.ndarray):
+            return self.solver.solve(b)
+        else:
+            assert False, f"The parameter type: {type(b)} is not support in linear solver for now."
```
Comment (yuanming-hu) [inline python/taichi/lang/sparse_solver.py:46]:
```suggestion
            assert False, f"The parameter type: {type(b)} is not supported in linear solvers for now."
```

File: python/taichi/lang/sparse_solver.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+            from taichi.lang.impl import get_runtime
+            taichi_arch = get_runtime().prog.config.arch
+            assert taichi_arch == _ti_core.Arch.x64 or taichi_arch == _ti_core.Arch.arm64, "SparseSolver only supports CPU for now."
+            self.solver = _ti_core.get_sparse_solver(solver_type)
+        else:
+            assert False, f"The solver type {solver_type} is not support for now. Only {solver_type_list} are supported."
+
+    @staticmethod
+    def type_assert(sparse_matrix):
+        assert False, f"The parameter type: {type(sparse_matrix)} is not support in linear solver for now."
```
Comment (yuanming-hu) [inline python/taichi/lang/sparse_solver.py:18]:
```suggestion
        assert False, f"The parameter type: {type(sparse_matrix)} is not supported in linear solvers for now."
```

Comment (yuanming-hu) [inline python/taichi/lang/sparse_solver.py:18]:
```suggestion
        assert False, f"The parameter type {type(sparse_matrix)} is not supported in linear solvers for now."
```

Comment (yuanming-hu) [inline python/taichi/lang/sparse_solver.py:18]:
```suggestion
        assert False, f"The parameter type: {type(sparse_matrix)} is not supported in linear solvers for now."
```

**Review by yuanming-hu — APPROVED** [review]
Thanks! LGTM. Please feel free to merge after these minor issues are fixed.

**Pull request merged**